### ISSUE TITLE
sql: separate different notions of "impure" for built-ins and var exprs 

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#26582]
Currently CockroachDB only knows two things: either "pure"/"const" or impure/var.
The former is always constant folded, the latter is never, but also never used to generate (e.g.) KV lookup spans.

This creates problems with e.g. `now()` which is not constant-foldable during a stmt prepare, but should be valid for generating spans.

The solution is to be found in pg's source code:

```c
/*
 * Symbolic values for provolatile column: these indicate whether the result
 * of a function is dependent *only* on the values of its explicit arguments,
 * or can change due to outside factors (such as parameter variables or
 * table contents).  NOTE: functions having side-effects, such as setval(),
 * must be labeled volatile to ensure they will not get optimized away,
 * even if the actual return value is not changeable.
 */
#define PROVOLATILE_IMMUTABLE   'i' /* never changes for given input */
#define PROVOLATILE_STABLE              's' /* does not change within a scan */
#define PROVOLATILE_VOLATILE    'v' /* can change even within a scan */
```

CockroachDB should annotate expressions as either immutable, stable or volatile. The planner can use non-folded stable expressions to generate spans.

cc @RaduBerinde

Informs #18836.

**Comment by justinj** [comment]
Bob and I talked about this a bit, my position is that this work would definitely be valuable to be done (it will certainly need to be done eventually), there's no pressing urgency for it right away since the plan cache machinery will probably need to be built out in more detail before we can make use of this information safely. It's definitely something we should keep in mind for the plan cache, though (cc @RaduBerinde @andy-kimball).

**Comment by knz** [comment]
Ok are you recommending that we postpone this work for now, and that Bob instead focuses on some area? (I don't mind either way -- I just need to know to organize the work).

**Comment by justinj** [comment]
I guess I can leave prioritization up to you, it's just my opinion that this work doesn't need to be done urgently (and won't be able to be utilized for a while in the future).

**Comment by knz** [comment]
thanks

**Comment by otan** [comment]
We'll need to do this for operators as well. For example, comparing timestamp and timestamptz is not immutable as it depends on session time zone.

**Comment by RaduBerinde** [comment]
Yeah, we currently special case division in the optimizer (it can error out).

**Comment by andy-kimball** [comment]
I _think_ we should be able to ignore dependencies on session time zone and other session state. We always invalidate plans when certain sensitive session state changes (though we may not be checking everything we need to). Other examples of cases where we invalidate plans are when session props like `ReorderJoinsLimit` and `ZigzagJoinEnabled` change. I think doing blanket invalidations on changes to sensitive session props is better than marking every sensitive function and operator as `stable`. That sounds harder and more fragile (and will make our plans slower). Session props rarely change, unlike `now`, and should not affect immutability.
